updated refarch chapter using supervisor domains terms (#31)
riscv-non-isa/riscv-ap-tee: specification/refarch.adoc

be multiples of the architectural page sizes supported by the hart MMU.
Similarly, the IOMMU via the IOMTT is expected to support a similar memory
tracking lookup to enable direct IOMMU and IOMMU-managed device access into
TVM confidential memory regions. For the CoVE reference architecture this TCB
consists of the HW (MMU, IOMMU, Memory Controller) and the SW/FW elements -
TSM-driver and the TSM. The TSM-driver is responsible for enforcing isolation of
confidential memory regions (consisting of multiple pages via MTT) and the TSM
is respondible for enforcing isolation of confidential memory pages among TVMs
(via G-stage translation) - pages assigned to the TVM may be exclusively
accessible to the condidential supervisor domain or may be shared with the
hosting supervisor domain (e.g. to allow for paravirtualized IO access).

[NOTE]
====
The TSM may manage additional attributes on TVM-assigned pages such as:
TVM-owner, Page-sub-type, TLB versioning information, Locking semaphore and
additional metadata etc.This extended memory tracking information managed by the
TSM is referred to as the Extended Memory Tracking Table (EMTT).
====

> ==== Handling Implicit, Explicit Accesses

Reviewer: I find this section title a bit ambiguous. What about:

```suggestion
==== Handling Implicit Memory Accesses
```

Author: updated to "Handling Implicit & Explicit Memory Accesses"

For TVM accesses for instruction fetch and page walks (virtual address
translation), isolated/confidential memory is required to enforce the
following security properties:

* TEE Instruction fetch - security property: TVM/TSM must not fetch code

> from untrusted/shared memory

Reviewer: Should we be explicit about the fact that this is enforced by the MMU?

> * TEE Instruction fetch - security property: TVM/TSM must not fetch code
> from untrusted/shared memory

Author: ```suggestion
* TEE Instruction fetch - security property: TVM/TSM must not fetch code
from untrusted/shared memory (this property must be enforced by the hart e.g. MMU)
```

=== TSM operation and properties

The TSM implements security routines that are invoked by the OS/VMM or by
the TVMs, e.g. by the VMM to grant a TVM a TEE-capable memory page and
setup second-stage mapping, activate a TVM virtual hart on a physical hart
etc. The TSM security routines are invoked by the OS/VMM via an ECALL with
the service call specified via registers. These service calls trap to the
TSM-driver. The TSM-driver switches hart state to the TSM context by
loading the hart's TSM execution state from the THCS.tssa and then returns
via an MRET to the TSM. The TSM executes the security routine requested
(where the TSM enforces the security properties) and may either return to

> the OS/VMM via an ECALL to the TSM-driver (TEERET with reason), or may use

Reviewer: ```suggestion
the hosting supervisor domain manager OS/VMM via an ECALL to the TSM-driver (TEERET with reason), or may use
```

In general, should we replace the `OS/VMM` instances with either `hosting supervisor domain manager` or `hosting supervisor domain manager OS/VMM` ?

Author: may be too verbose - will mention it at the intro and then stick with OS/VMM.

The TSM may not need to program stimecmp on its own, though it may verify
that time is not going back for a TVM. If the TSM needs to start a timer,
it should context switch the stimecmp CSR and replace it with its timeout
value if it's later than the timer it wants to start. The TSM may still
want to be aware of the value programmed into stimecmp to guard against
step attacks on TVMs.

Any NMIs experienced during TSM/TVM execution are always handled by the
TSM-driver and must cause the TEEs to be destroyed (preventing any loss of
confidential info via clearing of machine state). The TSM and therefore all
TVMs are prevented from execution after that point.

=== TSM and TVM Isolation

TSM (and all TVMs) memory is granted by the host OS/VMM but is isolated
(via access-control and/or confidentiality-protection) by the HW and TCB
elements. The TSM, TVM and HW isolation methods used must be evident in the
attestation evidence provided for the TVM since it identifies the hardware
and the TSM-driver.

There are two facets of TVM and TSM memory isolation that are
implementation-specific:

*a)* *Isolation from host software access* -  The CPU must enforce

> hardware-based access-control of TSM memory to prevent access from host software
> (VMM and host OS) V=0, HS-mode untrusted code (the hosting supervisor domain).

Reviewer: ```suggestion
hardware-based access-control of TSM memory to prevent access from host supervisor domain components
(VMM and host OS) V=0, HS-mode untrusted code.
```

exposed in volatile memory during execution. The use of an MEE and the
number of encryption domains supported is implementation-specific. For
example, The hardware may use the Supervisor Domain Identifier during execution
(and memory access) to cryptographically isolate memory associated with a
TEE which may be encrypted and additionally cryptographically
integrity-protected using a MAC on the memory contents. The MAC may be
maintained at various granularity - e.g. cache block size or in multiples
of cache blocks.

*TVM isolation* is the responsibility of the TSM via the G-stage
address translation table (hgatp). The TSM must track memory assignment of
TVMs (by the untrusted VMM/OS) to ensure memory assignment is
non-overlapping, along with additional security requirements. The security
requirements/invariants for enforcement of the memory
access-control for memory assigned to the TVMs is described in <<TVM Memory
management>>.

=== TVM Execution

As described above, TVMs can access both classes of memory - isolated memory
- which has confidentiality and access-control properties for memory exclusive
to the TVM, and non-confidential memory which is memory accessible to the host

> OS/VMM and is used for untrusted operations (e.g. virt-io, grpc communication

Reviewer: ```suggestion
OS/VMM and is used for untrusted operations (e.g. virtio, gRPC communication
```
